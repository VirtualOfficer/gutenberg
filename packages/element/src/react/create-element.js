/**
 * External dependencies
 */
import { createElement } from 'react';

/**
 * Returns a new element of given type. Type can be either a string tag name or
 * another function which itself returns an element.
 *
 * @param {?(string|Function)} type     Tag name or element creator.
 * @param {Object}             props    Element properties, either attribute
 *                                       set to apply to node or values to
 *                                       pass through to element creator.
 * @param {...WPElement}       children Descendant elements.
 *
 * @return {WPElement} Element.
 */
export default createElement;
