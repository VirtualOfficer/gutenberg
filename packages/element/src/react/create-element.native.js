/**
 * External dependencies
 */
import { get, isString } from 'lodash';
import { createElement as createElementDefault } from 'react';
import {
	Circle,
	Defs,
	Ellipse,
	G,
	Line,
	Path,
	Polygon,
	TextPath,
	TSpan,
	Svg,
} from 'react-native-svg';

const htmlTagToNativeComponent = {
	circle: Circle,
	defs: Defs,
	ellipse: Ellipse,
	g: G,
	line: Line,
	path: Path,
	polygon: Polygon,
	svg: Svg,
	textPath: TextPath,
	tspan: TSpan,
	/*Polyline,
	Text,
	Rect,
	Use,
	Image,
	Symbol,
	LinearGradient,
	RadialGradient,
	Stop,
	ClipPath*/
};

const mapHTMLTagToNativeComponent = ( type ) => {
	if ( ! isString( type ) ) {
		return type;
	}

	return get( htmlTagToNativeComponent, [ type ], type );
};

/**
 * Returns a new element of given type. Type can be either a string tag name or
 * another function which itself returns an element.
 *
 * @param {?(string|Function)} type     Tag name or element creator.
 * @param {Object}             props    Element properties, either attribute
 *                                       set to apply to node or values to
 *                                       pass through to element creator.
 * @param {...WPElement}       children Descendant elements.
 *
 * @return {WPElement} Element.
 */
const createElement = ( type, props, ...children ) => {
	return createElementDefault(
		mapHTMLTagToNativeComponent( type ),
		props,
		...children
	);
};

export default createElement;
