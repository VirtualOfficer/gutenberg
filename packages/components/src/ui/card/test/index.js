/**
 * External dependencies
 */
import { render } from '@testing-library/react';

/**
 * Internal dependencies
 */
import {
	Card,
	CardBody,
	CardFooter,
	CardHeader,
	CardInnerBody,
} from '../index';

describe( 'props', () => {
	test( 'should render correctly', () => {
		const { container } = render(
			<Card>
				<CardHeader title="WordPress.org" />
				<CardBody>Code is Poetry</CardBody>
				<CardFooter>
					<button>Agree</button>
				</CardFooter>
			</Card>
		);
		expect( container.firstChild ).toMatchSnapshot();
	} );

	test( 'should render elevation', () => {
		const { container } = render( <Card elevation /> );
		expect( container.firstChild ).toMatchSnapshot();
	} );

	test( 'should render no elevation', () => {
		const { container } = render( <Card elevation={ false } /> );
		expect( container.firstChild ).toMatchSnapshot();
	} );

	test( 'should render CardInnerBody', () => {
		const { container } = render(
			<Card>
				<CardHeader title="WordPress.org" />
				<CardInnerBody>Code is Poetry.</CardInnerBody>
				<CardFooter>
					<button>Agree</button>
				</CardFooter>
			</Card>
		);
		expect( container.firstChild ).toMatchSnapshot();
	} );
} );
