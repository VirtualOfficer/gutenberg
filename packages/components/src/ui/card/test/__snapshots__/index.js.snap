// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`props should render CardInnerBody 1`] = `
Snapshot Diff:
- First value
+ Second value

@@ -11,13 +11,15 @@
        data-wp-c16t="true"
        data-wp-component="CardHeader"
        title="WordPress.org"
      />
      <div
-       class="css-1maos9l components-card-inner-body css-1mm2cvy-View egi4jkx0"
+       class="components-scrollable wp-components-scrollable components-card-body ic-y3p1xi css-yk69wq css-1mm2cvy-View egi4jkx0"
+       data-g2-c16t="true"
+       data-g2-component="Scrollable"
        data-wp-c16t="true"
-       data-wp-component="CardInnerBody"
+       data-wp-component="CardBody"
      >
        Code is Poetry.
      </div>
      <div
        class="components-flex css-1tfkqy8 css-196clky css-0 components-card-footer css-138amlm-Flex-ItemsRow css-1mm2cvy-View egi4jkx0"
`;

exports[`props should render correctly 1`] = `
.emotion-0.emotion-0.emotion-0.emotion-0.emotion-0.emotion-0.emotion-0 {
  border-bottom: 1px solid;
}

.emotion-2.emotion-2.emotion-2.emotion-2.emotion-2.emotion-2.emotion-2 {
  border-color: rgba(0, 0, 0, 0.1);
  border-color: var(--wp-g2-color-divider);
  min-height: 44px;
  min-height: var(--wp-g2-card-header-height);
  padding-bottom: calc(4px * 1);
  padding-bottom: var(--wp-g2-card-header-footer-padding-y);
  padding-left: calc(4px * 3);
  padding-left: var(--wp-g2-card-padding-x);
  padding-right: calc(4px * 3);
  padding-right: var(--wp-g2-card-padding-x);
  padding-top: calc(4px * 1);
  padding-top: var(--wp-g2-card-header-footer-padding-y);
}

.emotion-1.emotion-1.emotion-1.emotion-1.emotion-1.emotion-1.emotion-1:first-of-type {
  border-top-left-radius: 2px;
  border-top-left-radius: var(--wp-g2-card-border-radius);
  border-top-right-radius: 2px;
  border-top-right-radius: var(--wp-g2-card-border-radius);
}

.emotion-1.emotion-1.emotion-1.emotion-1.emotion-1.emotion-1.emotion-1:last-of-type {
  border-bottom-left-radius: 2px;
  border-bottom-left-radius: var(--wp-g2-card-border-radius);
  border-bottom-right-radius: 2px;
  border-bottom-right-radius: var(--wp-g2-card-border-radius);
}

.emotion-28.emotion-28.emotion-28.emotion-28.emotion-28.emotion-28.emotion-28 {
  background-color: #ffffff;
  background-color: var(--wp-g2-surface-color);
  color: #1e1e1e;
  color: var(--wp-g2-color-text);
  position: relative;
  --wp-g2-surface-background-size: 12px;
  --wp-g2-surface-background-size-dotted: 11px;
  box-shadow: 0 0 0 1px rgba(0, 0, 0, 0.1);
  box-shadow: 0 0 0 1px var(--wp-g2-surface-border-color);
  outline: none;
  border-radius: 2px;
  border-radius: var(--wp-g2-card-border-radius);
}

.emotion-7.emotion-7.emotion-7.emotion-7.emotion-7.emotion-7.emotion-7 {
  height: 100%;
  overflow-x: hidden;
  overflow-y: auto;
  height: auto;
  max-height: 100%;
  padding: calc(4px * 3) calc(4px * 3);
  padding: var(--wp-g2-card-padding);
}

@media only screen and ( min-device-width:40em ) {
  .emotion-7.emotion-7.emotion-7.emotion-7.emotion-7.emotion-7.emotion-7::-webkit-scrollbar {
    height: 12px;
    width: 12px;
  }

  .emotion-7.emotion-7.emotion-7.emotion-7.emotion-7.emotion-7.emotion-7::-webkit-scrollbar-track {
    background-color: transparent;
  }

  .emotion-7.emotion-7.emotion-7.emotion-7.emotion-7.emotion-7.emotion-7::-webkit-scrollbar-track {
    background: rgba(0, 0, 0, 0.04);
    background: var(--wp-g2-color-scrollbar-track);
    border-radius: 8px;
  }

  .emotion-7.emotion-7.emotion-7.emotion-7.emotion-7.emotion-7.emotion-7::-webkit-scrollbar-thumb {
    background-clip: padding-box;
    background-color: rgba(0, 0, 0, 0.2);
    background-color: var(--wp-g2-color-scrollbar-thumb);
    border: 2px solid rgba( 0,0,0,0 );
    border-radius: 7px;
  }

  .emotion-7.emotion-7.emotion-7.emotion-7.emotion-7.emotion-7.emotion-7:hover::-webkit-scrollbar-thumb {
    background-color: rgba(0, 0, 0, 0.5);
    background-color: var(--wp-g2-color-scrollbar-thumb-hover);
  }
}

.emotion-7.emotion-7.emotion-7.emotion-7.emotion-7.emotion-7.emotion-7:first-of-type {
  border-top-left-radius: 2px;
  border-top-left-radius: var(--wp-g2-card-border-radius);
  border-top-right-radius: 2px;
  border-top-right-radius: var(--wp-g2-card-border-radius);
}

.emotion-7.emotion-7.emotion-7.emotion-7.emotion-7.emotion-7.emotion-7:last-of-type {
  border-bottom-left-radius: 2px;
  border-bottom-left-radius: var(--wp-g2-card-border-radius);
  border-bottom-right-radius: 2px;
  border-bottom-right-radius: var(--wp-g2-card-border-radius);
}

.emotion-19.emotion-19.emotion-19.emotion-19.emotion-19.emotion-19.emotion-19 {
  border-radius: inherit;
  bottom: 0;
  box-shadow: 0 1px 2px 0 rgba(0 ,0,0,0.05);
  opacity: 1;
  opacity: var(--wp-g2-elevation-intensity);
  left: 0;
  right: 0;
  top: 0;
  -webkit-transition: box-shadow 200ms cubic-bezier(0.08, 0.52, 0.52, 1);
  -webkit-transition: box-shadow var(--wp-g2-transition-duration) var(--wp-g2-transition-timing-function);
  transition: box-shadow 200ms cubic-bezier(0.08, 0.52, 0.52, 1);
  transition: box-shadow var(--wp-g2-transition-duration) var(--wp-g2-transition-timing-function);
}

.emotion-24.emotion-24.emotion-24.emotion-24.emotion-24.emotion-24.emotion-24 {
  border-radius: inherit;
  bottom: 0;
  box-shadow: 0 2px 4px 0 rgba(0 ,0,0,0.1);
  opacity: 1;
  opacity: var(--wp-g2-elevation-intensity);
  left: 0;
  right: 0;
  top: 0;
  -webkit-transition: box-shadow 200ms cubic-bezier(0.08, 0.52, 0.52, 1);
  -webkit-transition: box-shadow var(--wp-g2-transition-duration) var(--wp-g2-transition-timing-function);
  transition: box-shadow 200ms cubic-bezier(0.08, 0.52, 0.52, 1);
  transition: box-shadow var(--wp-g2-transition-duration) var(--wp-g2-transition-timing-function);
}

.emotion-18 {
  background: transparent;
  display: block;
  margin: 0 !important;
  pointer-events: none;
  position: absolute;
  will-change: box-shadow;
}

.emotion-20 {
  border-radius: 2px;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  width: 100%;
}

.emotion-4 > * + *:not(marquee) {
  margin-left: calc(4px * 2);
}

.emotion-4 > * {
  min-width: 0;
}

.emotion-13 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  width: 100%;
}

.emotion-13 > * + *:not(marquee) {
  margin-left: calc(4px * 2);
}

.emotion-13 > * {
  min-width: 0;
}

<div
  class="components-surface components-card emotion-28 emotion-5 emotion-6"
  data-wp-c16t="true"
  data-wp-component="Card"
>
  <div
    class="emotion-5 emotion-6"
  >
    <div
      class="components-flex emotion-0 emotion-1 emotion-2 emotion-3 components-card-header emotion-4 emotion-5 emotion-6"
      data-wp-c16t="true"
      data-wp-component="CardHeader"
      title="WordPress.org"
    />
    <div
      class="components-scrollable wp-components-scrollable components-card-body ic-y3p1xi emotion-7 emotion-5 emotion-6"
      data-g2-c16t="true"
      data-g2-component="Scrollable"
      data-wp-c16t="true"
      data-wp-component="CardBody"
    >
      Code is Poetry.
    </div>
    <div
      class="components-flex emotion-1 emotion-2 emotion-3 components-card-footer emotion-13 emotion-5 emotion-6"
      data-wp-c16t="true"
      data-wp-component="CardFooter"
    >
      <button>
        Agree
      </button>
    </div>
  </div>
  <div
    aria-hidden="true"
    class="emotion-18 emotion-19 components-elevation emotion-20 emotion-5 emotion-6"
    data-wp-c16t="true"
    data-wp-component="Elevation"
  />
  <div
    aria-hidden="true"
    class="emotion-18 emotion-24 components-elevation emotion-20 emotion-5 emotion-6"
    data-wp-c16t="true"
    data-wp-component="Elevation"
  />
</div>
`;

exports[`props should render elevation 1`] = `
Snapshot Diff:
- First value
+ Second value

@@ -6,16 +6,16 @@
    <div
      class="css-1mm2cvy-View egi4jkx0"
    />
    <div
      aria-hidden="true"
-     class="css-zsg5ag-Elevation css-1yc4u8p components-elevation css-icip60 css-1mm2cvy-View egi4jkx0"
+     class="css-zsg5ag-Elevation css-lxi9t4 components-elevation css-icip60 css-1mm2cvy-View egi4jkx0"
      data-wp-c16t="true"
      data-wp-component="Elevation"
    />
    <div
      aria-hidden="true"
-     class="css-zsg5ag-Elevation css-akhjqa components-elevation css-icip60 css-1mm2cvy-View egi4jkx0"
+     class="css-zsg5ag-Elevation css-19rc1q1 components-elevation css-icip60 css-1mm2cvy-View egi4jkx0"
      data-wp-c16t="true"
      data-wp-component="Elevation"
    />
  </div>
`;
